api/lib/opentelemetry/trace/tracer.rb

```diff
@@ -0,0 +1,91 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # No-op implementations of Tracer.
+    module Tracer
+      extend self
+
+      CONTEXT_SPAN_KEY = :__span__
+      HTTP_TEXT_FORMAT = nil # TODO: implement HttpTraceContext
+
+      # Formatter for serializing and deserializing a SpanContext into a binary format.
+      module NoopBinaryFormat
+        extend self
+
+        def to_bytes(span_context)
+          raise ArgumentError if span_context.nil?
+
+          []
+        end
+
+        def from_bytes(bytes)
+          raise ArgumentError if bytes.nil?
+
+          SpanContext.INVALID
+        end
+      end
+
+      private_constant(:CONTEXT_SPAN_KEY, :HTTP_TEXT_FORMAT, :NoopBinaryFormat)
+
+      def current_span
+        Context.get(CONTEXT_SPAN_KEY) || Span.INVALID
+      end
+
+      # TODO: This is a helper for the default use-case of extending the current trace with a span.
+      #
+      # The spec-ed API seems a little clunky. Default use-case looks like:
+      # OpenTelemetry.tracer.with_span(OpenTelemetry.tracer.start_span('do-the-thing')) do ... end
+      #
+      # OpenTracing equivalent looks like:
+      # OpenTracing.start_active_span('do-the-thing') do ... end
+      #
+      # OpenCensus equivalent looks like:
+      # OpenCensus::Trace.in_span('do-the-thing') do ... end
+      #
+      # With this helper:
+      # OpenTelemetry.tracer.in_span('do-the-thing') do ... end
+      def in_span(name, sampler: nil, links: nil, recording_events: nil, kind: nil)
+        span = start_span(name, sampler: sampler, links: links, recording_events: recording_events, kind: kind)
+        with_span(span) { |s| yield s }
+      end
+
+      def with_span(span)
+        Context.with(CONTEXT_SPAN_KEY, span) { |s| yield s }
+      end
+
+      def start_root_span(name, sampler: nil, links: nil, recording_events: nil, kind: nil)
+        raise ArgumentError if name.nil?
```

Comment:
I'm not sure what we should do for invalid arguments in the minimal implementation, and somewhat torn on the issue.

I just wanted to point out that there is a proposal currently in PR form here: https://github.com/open-telemetry/opentelemetry-specification/pull/153.

There seem to be various opinions as you read through the comments. This [comment](https://github.com/open-telemetry/opentelemetry-specification/pull/153#discussion_r296447000) proposed a somewhat pragmatic approach of falling back to a default when an argument is invalid. Quote (from the comment)

> Specific case: I do think using a null name for Span could fallback to using a default name, instead of throwing. Documenting it is nice, but providing defaults for these kind of things is a good alternative ;)

It might be too early to make a final call on how to deal exceptions in the minimal implementation, but we should keep an eye on that PR and the specification repo for guidance.